File: db-migrate-dotnet/Services/TargetWriter.cs
using Fuxa.DbMigrate.Models;
using SqlSugar;

namespace Fuxa.DbMigrate.Services;

/// <summary>
/// Writes data to MySQL or SQL Server using SqlSugar ORM
/// </summary>
public class TargetWriter : IDisposable
{
    private readonly TargetConfig _config;
    private SqlSugarClient? _db;
    private bool _disposed;

    public TargetWriter(TargetConfig config)
    {
        _config = config;
    }

    private bool IsMySql => _config.DbType.Equals("MySql", StringComparison.OrdinalIgnoreCase);

    /// <summary>
    /// Initialize the SqlSugar connection
    /// </summary>
    public void Connect()
    {
        var connStr = BuildConnectionString();
        var dbType = IsMySql ? DbType.MySql : DbType.SqlServer;

        _db = new SqlSugarClient(new ConnectionConfig
        {
            ConnectionString = connStr,
            DbType = dbType,
            IsAutoCloseConnection = true,
        });
    }

    /// <summary>
    /// Test the database connection
    /// </summary>
    public async Task<bool> TestConnectionAsync()
    {
        EnsureConnected();
        var result = await _db!.Ado.GetIntAsync("SELECT 1");
        return result == 1;
    }

    /// <summary>
    /// Create a table in the target database, optionally dropping it first
    /// </summary>
    public async Task CreateTableAsync(string tableName, TableDef tableDef, string? dbPrefix = null, bool dropIfExists = false)
    {
        EnsureConnected();
        var fullName = dbPrefix != null ? $"{dbPrefix}_{tableName}" : tableName;

        if (dropIfExists)
        {
            await DropTableAsync(fullName);
        }

        var sql = IsMySql ? BuildCreateTableMySql(fullName, tableDef) : BuildCreateTableMssql(fullName, tableDef);
        await _db!.Ado.ExecuteCommandAsync(sql);
    }

    /// <summary>
    /// Drop a table if it exists
    /// </summary>
    private async Task DropTableAsync(string tableName)
    {
        EnsureConnected();
        if (IsMySql)
        {
            await _db!.Ado.ExecuteCommandAsync($"DROP TABLE IF EXISTS `{tableName}`");
        }
        else
        {
            await _db!.Ado.ExecuteCommandAsync(
                $"IF OBJECT_ID('dbo.[{tableName}]', 'U') IS NOT NULL DROP TABLE dbo.[{tableName}]");
        }
    }

    /// <summary>
    /// Insert rows into a table using batch insert (UPSERT when primary key exists)
    /// </summary>
    public async Task<long> InsertRowsAsync(string tableName, List<Dictionary<string, object?>> rows,
        TableDef tableDef, string? dbPrefix = null, int batchSize = 1000)
    {
        if (rows.Count == 0) return 0;

        EnsureConnected();
        var fullName = dbPrefix != null ? $"{dbPrefix}_{tableName}" : tableName;
        long totalInserted = 0;

        // Process in batches
        for (int i = 0; i < rows.Count; i += batchSize)
        {
            var batch = rows.Skip(i).Take(batchSize).ToList();
            await InsertBatchAsync(fullName, batch, tableDef);
            totalInserted += batch.Count;
        }

        return totalInserted;
    }

    /// <summary>
    /// Insert a single batch of rows
    /// </summary>
    private async Task InsertBatchAsync(string tableName, List<Dictionary<string, object?>> rows, TableDef tableDef)
    {
        if (rows.Count == 0) return;

        var columns = tableDef.Columns;
        var colNames = string.Join(", ", columns.Select(c => IsMySql ? $"`{c.Name}`" : $"[{c.Name}]"));

        if (tableDef.PrimaryKey != null)
        {
            // UPSERT: use REPLACE INTO for MySQL, MERGE for SQL Server
            if (IsMySql)
            {
                await UpsertBatchMySqlAsync(tableName, rows, tableDef);
            }
            else
            {
                await UpsertBatchMssqlAsync(tableName, rows, tableDef);
            }
        }
        else
        {
            // Plain INSERT
            await PlainInsertBatchAsync(tableName, rows, tableDef);
        }
    }

    /// <summary>
    /// MySQL: REPLACE INTO for upsert
    /// </summary>
    private async Task UpsertBatchMySqlAsync(string tableName, List<Dictionary<string, object?>> rows, TableDef tableDef)
    {
        var colNames = string.Join(", ", tableDef.Columns.Select(c => $"`{c.Name}`"));
        var paramList = new List<string>();
        var parameters = new List<SugarParameter>();
        int paramIdx = 0;

        foreach (var row in rows)
        {
            var placeholders = new List<string>();
            foreach (var col in tableDef.Columns)
            {
                var paramName = $"@p{paramIdx++}";
                placeholders.Add(paramName);
                parameters.Add(new SugarParameter(paramName, FormatValue(row.GetValueOrDefault(col.Name), col)));
            }
            paramList.Add($"({string.Join(", ", placeholders)})");
        }

        var sql = $"REPLACE INTO `{tableName}` ({colNames}) VALUES {string.Join(", ", paramList)}";
        await _db!.Ado.ExecuteCommandAsync(sql, parameters.ToArray());
    }

    /// <summary>
    /// SQL Server: MERGE for upsert
    /// </summary>
    private async Task UpsertBatchMssqlAsync(string tableName, List<Dictionary<string, object?>> rows, TableDef tableDef)
    {
        var pk = tableDef.PrimaryKey!;
        var colNames = tableDef.Columns.Select(c => $"[{c.Name}]").ToList();
        var colNamesStr = string.Join(", ", colNames);
        var parameters = new List<SugarParameter>();
        int paramIdx = 0;

        // Build VALUES clause
        var valueRows = new List<string>();
        foreach (var row in rows)
        {
            var placeholders = new List<string>();
            foreach (var col in tableDef.Columns)
            {
                var paramName = $"@p{paramIdx++}";
                placeholders.Add(paramName);
                parameters.Add(new SugarParameter(paramName, FormatValue(row.GetValueOrDefault(col.Name), col)));
            }
            valueRows.Add($"({string.Join(", ", placeholders)})");
        }

        var valuesStr = string.Join(", ", valueRows);
        var sourceCols = string.Join(", ", tableDef.Columns.Select(c => $"src.[{c.Name}]"));
        var updateSet = string.Join(", ", tableDef.Columns
            .Where(c => c.Name != pk)
            .Select(c => $"tgt.[{c.Name}] = src.[{c.Name}]"));

        var sql = $"""
            MERGE INTO [{tableName}] AS tgt
            USING (VALUES {valuesStr}) AS src ({colNamesStr})
            ON tgt.[{pk}] = src.[{pk}]
            WHEN MATCHED THEN UPDATE SET {updateSet}
            WHEN NOT MATCHED THEN INSERT ({colNamesStr}) VALUES ({sourceCols});
            """;

        await _db!.Ado.ExecuteCommandAsync(sql, parameters.ToArray());
    }

    /// <summary>
    /// Plain INSERT (no primary key, no upsert)
    /// </summary>
    private async Task PlainInsertBatchAsync(string tableName, List<Dictionary<string, object?>> rows, TableDef tableDef)
    {
        var colNames = string.Join(", ", tableDef.Columns.Select(c => IsMySql ? $"`{c.Name}`" : $"[{c.Name}]"));
        var paramList = new List<string>();
        var parameters = new List<SugarParameter>();
        int paramIdx = 0;

        foreach (var row in rows)
        {
            var placeholders = new List<string>();
            foreach (var col in tableDef.Columns)
            {
                var paramName = $"@p{paramIdx++}";
                placeholders.Add(paramName);
                parameters.Add(new SugarParameter(paramName, FormatValue(row.GetValueOrDefault(col.Name), col)));
            }
            paramList.Add($"({string.Join(", ", placeholders)})");
        }

        var prefix = IsMySql ? $"`{tableName}`" : $"[{tableName}]";
        var sql = $"INSERT INTO {prefix} ({colNames}) VALUES {string.Join(", ", paramList)}";
        await _db!.Ado.ExecuteCommandAsync(sql, parameters.ToArray());
    }

    /// <summary>
    /// Format a value for insertion based on column type
    /// </summary>
    private static object? FormatValue(object? value, ColumnDef col)
    {
        if (value == null || value is DBNull)
            return DBNull.Value;

        // Integer types
        if (col.MysqlType is "INT" or "BIGINT" or "INT AUTO_INCREMENT")
        {
            if (long.TryParse(value.ToString(), out var num))
                return num;
            return DBNull.Value;
        }

        // DateTime types
        if (col.MysqlType is "DATETIME")
        {
            if (value is DateTime dt)
                return dt;
            if (DateTime.TryParse(value.ToString(), out var parsed))
                return parsed;
            return DBNull.Value;
        }

        return value.ToString();
    }

    #region DDL Builders

    private static string BuildCreateTableMySql(string tableName, TableDef tableDef)
    {
        var columns = tableDef.Columns.Select(col => $"  `{col.Name}` {col.MysqlType}");
        var pk = tableDef.PrimaryKey != null
            ? $", PRIMARY KEY (`{tableDef.PrimaryKey}`)"
            : "";

        return $"CREATE TABLE IF NOT EXISTS `{tableName}` (\n{string.Join(",\n", columns)}{pk}\n) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci";
    }

    private static string BuildCreateTableMssql(string tableName, TableDef tableDef)
    {
        var columns = tableDef.Columns.Select(col => $"  [{col.Name}] {col.MssqlType}");
        var pk = tableDef.PrimaryKey != null
            ? $", CONSTRAINT [PK_{tableName}] PRIMARY KEY ([{tableDef.PrimaryKey}])"
            : "";

        return $"""
            IF NOT EXISTS (SELECT * FROM sysobjects WHERE name='{tableName}' AND xtype='U')
            CREATE TABLE [{tableName}] (
            {string.Join(",\n", columns)}{pk}
            )
            """;
    }

    #endregion

    private string BuildConnectionString()
    {
        if (IsMySql)
        {
            return $"Server={_config.Host};Port={_config.Port};Database={_config.Database};Uid={_config.User};Pwd={_config.Password};CharSet=utf8mb4;AllowUserVariables=True;";
        }
        else
        {
            var encrypt = _config.Encrypt ? "True" : "False";
            var trustCert = _config.TrustServerCertificate ? "True" : "False";
            return $"Server={_config.Host},{_config.Port};Database={_config.Database};User Id={_config.User};Password={_config.Password};Encrypt={encrypt};TrustServerCertificate={trustCert};";
        }
    }

    private void EnsureConnected()
    {
        if (_db == null)
            throw new InvalidOperationException("Not connected. Call Connect() first.");
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _db?.Dispose();
        GC.SuppressFinalize(this);
    }
}

File: db-migrate/src/target-writer.js
/**
 * Target database writer module
 * Supports writing to MySQL and SQL Server
 */

'use strict';

class TargetWriter {
  /**
   * @param {object} config - target database config { type: 'mysql'|'mssql', ... }
   */
  constructor(config) {
    this.config = config;
    this.type = config.type; // 'mysql' or 'mssql'
    this.connection = null;
  }

  /**
   * Connect to the target database
   */
  async connect() {
    if (this.type === 'mysql') {
      return this._connectMysql();
    } else if (this.type === 'mssql') {
      return this._connectMssql();
    }
    throw new Error(`Unsupported target type: ${this.type}`);
  }

  async _connectMysql() {
    const mysql = require('mysql2/promise');
    const conn = await mysql.createConnection({
      host: this.config.host || 'localhost',
      port: this.config.port || 3306,
      user: this.config.user,
      password: this.config.password,
      database: this.config.database,
      multipleStatements: true,
    });
    this.connection = conn;
    return conn;
  }

  async _connectMssql() {
    const sql = require('mssql');
    const pool = await sql.connect({
      server: this.config.host || 'localhost',
      port: this.config.port || 1433,
      user: this.config.user,
      password: this.config.password,
      database: this.config.database,
      options: {
        encrypt: this.config.encrypt || false,
        trustServerCertificate: this.config.trustServerCertificate || true,
      },
      requestTimeout: this.config.requestTimeout || 60000,
      pool: { max: 10, min: 0, idleTimeoutMillis: 30000 },
    });
    this.connection = pool;
    return pool;
  }

  /**
   * Drop a table if it exists
   * @param {string} tableName
   * @param {string} dbPrefix
   */
  async dropTable(tableName, dbPrefix) {
    const fullTableName = dbPrefix ? `${dbPrefix}_${tableName}` : tableName;

    if (this.type === 'mysql') {
      await this.connection.query(`DROP TABLE IF EXISTS \`${fullTableName}\``);
    } else {
      try {
        await this.connection.request().query(`DROP TABLE IF EXISTS [${fullTableName}]`);
      } catch (err) {
        // ignore
      }
    }
  }

  /**
   * Create a table in the target database
   * @param {string} tableName
   * @param {object} tableDef - table definition from schema.js
   * @param {string} dbPrefix - optional prefix for table name (e.g. for DAQ databases)
   * @param {boolean} dropIfExists - drop the table first if it exists
   */
  async createTable(tableName, tableDef, dbPrefix, dropIfExists = false) {
    const fullTableName = dbPrefix ? `${dbPrefix}_${tableName}` : tableName;

    if (dropIfExists) {
      await this.dropTable(tableName, dbPrefix);
    }

    if (this.type === 'mysql') {
      return this._createTableMysql(fullTableName, tableDef);
    } else {
      return this._createTableMssql(fullTableName, tableDef);
    }
  }

  async _createTableMysql(tableName, tableDef) {
    const columns = tableDef.columns.map(col => {
      let def = `\`${col.name}\` ${col.mysqlType}`;
      // Don't add NOT NULL for nullable columns
      return def;
    });

    let pk = '';
    if (tableDef.primaryKey && !tableDef.autoIncrement) {
      pk = `, PRIMARY KEY (\`${tableDef.primaryKey}\`)`;
    } else if (tableDef.primaryKey && tableDef.autoIncrement) {
      // auto increment column is already primary key
      pk = `, PRIMARY KEY (\`${tableDef.primaryKey}\`)`;
    }

    const sql = `CREATE TABLE IF NOT EXISTS \`${tableName}\` (\n  ${columns.join(',\n  ')}${pk}\n) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci`;

    await this.connection.query(sql);
  }

  async _createTableMssql(tableName, tableDef) {
    const columns = tableDef.columns.map(col => {
      let def = `[${col.name}] ${col.mssqlType}`;
      return def;
    });

    let pk = '';
    if (tableDef.primaryKey) {
      pk = `, CONSTRAINT [PK_${tableName}] PRIMARY KEY ([${tableDef.primaryKey}])`;
    }

    const sql = `IF NOT EXISTS (SELECT * FROM sysobjects WHERE name='${tableName}' AND xtype='U')\nCREATE TABLE [${tableName}] (\n  ${columns.join(',\n  ')}${pk}\n)`;

    await this.connection.request().query(sql);
  }

  /**
   * Insert rows into a table (UPSERT)
   * @param {string} tableName
   * @param {Array} rows
   * @param {object} tableDef
   * @param {string} dbPrefix
   * @param {number} batchSize
   */
  async insertRows(tableName, rows, tableDef, dbPrefix, batchSize = 500) {
    if (!rows || rows.length === 0) return 0;

    const fullTableName = dbPrefix ? `${dbPrefix}_${tableName}` : tableName;
    let totalInserted = 0;

    // Process in batches
    for (let i = 0; i < rows.length; i += batchSize) {
      const batch = rows.slice(i, i + batchSize);

      if (this.type === 'mysql') {
        await this._insertBatchMysql(fullTableName, batch, tableDef);
      } else {
        await this._insertBatchMssql(fullTableName, batch, tableDef);
      }
      totalInserted += batch.length;
    }

    return totalInserted;
  }

  async _insertBatchMysql(tableName, rows, tableDef) {
    if (rows.length === 0) return;

    const columns = tableDef.columns.map(c => `\`${c.name}\``).join(', ');
    const placeholders = tableDef.columns.map(() => '?').join(', ');
    const valuesClause = rows.map(() => `(${placeholders})`).join(', ');

    // Use REPLACE INTO for upsert (requires primary key)
    let sql;
    if (tableDef.primaryKey) {
      sql = `REPLACE INTO \`${tableName}\` (${columns}) VALUES ${valuesClause}`;
    } else {
      sql = `INSERT INTO \`${tableName}\` (${columns}) VALUES ${valuesClause}`;
    }

    const flatValues = [];
    for (const row of rows) {
      for (const col of tableDef.columns) {
        flatValues.push(this._formatValue(row[col.name], col));
      }
    }

    await this.connection.query(sql, flatValues);
  }

  async _insertBatchMssql(tableName, rows, tableDef) {
    if (rows.length === 0) return;

    const sql = require('mssql');
    const table = new sql.Table(tableName);
    table.create = false;

    for (const col of tableDef.columns) {
      const typeInfo = this._getMssqlType(col);
      table.columns.add(col.name, typeInfo.type, { nullable: true, ...typeInfo.options });
    }

    for (const row of rows) {
      const values = tableDef.columns.map(col => this._formatValue(row[col.name], col));
      table.rows.add(...values);
    }

    const request = this.connection.request();
    await request.bulk(table);
  }

  _getMssqlType(col) {
    const sql = require('mssql');
    // BCP bulk insert requires the type constructor (class), NOT an instance.
    // For variable-length types, pass length via options.
    const typeMap = {
      'INT': { type: sql.Int },
      'INT IDENTITY(1,1)': { type: sql.Int },
      'INT AUTO_INCREMENT': { type: sql.Int },
      'BIGINT': { type: sql.BigInt },
      'VARCHAR(100)': { type: sql.NVarChar, options: { length: 100 } },
      'VARCHAR(255)': { type: sql.NVarChar, options: { length: 255 } },
      'NVARCHAR(100)': { type: sql.NVarChar, options: { length: 100 } },
      'NVARCHAR(255)': { type: sql.NVarChar, options: { length: 255 } },
      'NVARCHAR(MAX)': { type: sql.NVarChar, options: { length: 'max' } },
      'TEXT': { type: sql.NVarChar, options: { length: 'max' } },
      'DATETIME': { type: sql.DateTime2 },
      'DATETIME2': { type: sql.DateTime2 },
    };
    return typeMap[col.mssqlType] || { type: sql.NVarChar, options: { length: 'max' } };
  }

  /**
   * Format value for insertion
   */
  _formatValue(value, col) {
    if (value === undefined || value === null) {
      return null;
    }
    // Integer types
    if (col.mysqlType === 'INT' || col.mysqlType === 'BIGINT' ||
        col.mysqlType === 'INT AUTO_INCREMENT') {
      const num = parseInt(value);
      return isNaN(num) ? null : num;
    }
    return value;
  }

  /**
   * Truncate a table before migration
   * @param {string} tableName
   * @param {string} dbPrefix
   */
  async truncateTable(tableName, dbPrefix) {
    const fullTableName = dbPrefix ? `${dbPrefix}_${tableName}` : tableName;

    if (this.type === 'mysql') {
      try {
        await this.connection.query(`SET FOREIGN_KEY_CHECKS = 0`);
        await this.connection.query(`TRUNCATE TABLE \`${fullTableName}\``);
        await this.connection.query(`SET FOREIGN_KEY_CHECKS = 1`);
      } catch (err) {
        // Table might not exist
        if (!err.message.includes("doesn't exist")) {
          throw err;
        }
      }
    } else {
      try {
        await this.connection.request().query(`TRUNCATE TABLE [${fullTableName}]`);
      } catch (err) {
        // Table might not exist
        if (!err.message.includes('Cannot find')) {
          throw err;
        }
      }
    }
  }

  /**
   * Test connection
   * @returns {boolean}
   */
  async testConnection() {
    try {
      if (this.type === 'mysql') {
        const [rows] = await this.connection.query('SELECT 1 as test');
        return rows && rows.length > 0;
      } else {
        const result = await this.connection.request().query('SELECT 1 as test');
        return result && result.recordset && result.recordset.length > 0;
      }
    } catch (err) {
      throw err;
    }
  }

  /**
   * Close the connection
   */
  async close() {
    if (this.connection) {
      if (this.type === 'mysql') {
        await this.connection.end();
      } else {
        await this.connection.close();
      }
      this.connection = null;
    }
  }
}

module.exports = TargetWriter;
